File: frontend/src/api/scans.ts
import { api } from './client'
import type { ScanConfig, ScanConfigPreview, ScanJob } from '../types'

export const scansApi = {
  list: (slug: string) =>
    api.get<ScanConfig[]>(`/projects/${slug}/scans`),

  get: (slug: string, scanId: string) =>
    api.get<ScanConfig>(`/projects/${slug}/scans/${scanId}`),

  create: (slug: string, data: {
    data_source_id: string
    name: string
    base_query: string
    event_type_id?: string | null
    event_type_column?: string | null
    time_column?: string | null
    event_name_format?: string | null
    json_value_paths?: string[]
    cardinality_threshold?: number
    interval?: string | null
  }) => api.post<ScanConfig>(`/projects/${slug}/scans`, data),

  preview: (slug: string, data: {
    data_source_id: string
    base_query: string
    limit?: number
  }) => api.post<ScanConfigPreview>(`/projects/${slug}/scans/preview`, data),

  update: (slug: string, scanId: string, data: {
    name?: string
    base_query?: string
    event_type_id?: string | null
    event_type_column?: string | null
    time_column?: string | null
    event_name_format?: string | null
    json_value_paths?: string[]
    cardinality_threshold?: number
    interval?: string | null
  }) => api.patch<ScanConfig>(`/projects/${slug}/scans/${scanId}`, data),

  del: (slug: string, scanId: string) =>
    api.del(`/projects/${slug}/scans/${scanId}`),

  run: (slug: string, scanId: string) =>
    api.post<ScanJob>(`/projects/${slug}/scans/${scanId}/run`, {}),

  listJobs: (slug: string, scanId: string) =>
    api.get<ScanJob[]>(`/projects/${slug}/scans/${scanId}/jobs`),

  getJob: (slug: string, scanId: string, jobId: string) =>
    api.get<ScanJob>(`/projects/${slug}/scans/${scanId}/jobs/${jobId}`),
}

File: backend/src/tripl/api/v1/scans.py
import uuid

from fastapi import APIRouter

from tripl.api.deps import SessionDep
from tripl.schemas.scan_config import (
    ScanConfigCreate,
    ScanConfigPreviewRequest,
    ScanConfigPreviewResponse,
    ScanConfigResponse,
    ScanConfigUpdate,
)
from tripl.schemas.scan_job import ScanJobResponse
from tripl.services import scan_service

router = APIRouter(
    prefix="/projects/{slug}/scans",
    tags=["scans"],
)


@router.get("", response_model=list[ScanConfigResponse])
async def list_scan_configs(session: SessionDep, slug: str):
    return await scan_service.list_scan_configs(session, slug)


@router.post("", response_model=ScanConfigResponse, status_code=201)
async def create_scan_config(
    session: SessionDep, slug: str, data: ScanConfigCreate
):
    return await scan_service.create_scan_config(session, slug, data)


@router.post("/preview", response_model=ScanConfigPreviewResponse)
async def preview_scan_config(
    session: SessionDep,
    slug: str,
    data: ScanConfigPreviewRequest,
):
    return await scan_service.preview_scan_config(session, slug, data)


@router.get("/{scan_id}", response_model=ScanConfigResponse)
async def get_scan_config(session: SessionDep, slug: str, scan_id: uuid.UUID):
    return await scan_service.get_scan_config(session, slug, scan_id)


@router.patch("/{scan_id}", response_model=ScanConfigResponse)
async def update_scan_config(
    session: SessionDep,
    slug: str,
    scan_id: uuid.UUID,
    data: ScanConfigUpdate,
):
    return await scan_service.update_scan_config(session, slug, scan_id, data)


@router.delete("/{scan_id}", status_code=204)
async def delete_scan_config(session: SessionDep, slug: str, scan_id: uuid.UUID):
    await scan_service.delete_scan_config(session, slug, scan_id)


@router.post("/{scan_id}/run", response_model=ScanJobResponse, status_code=201)
async def run_scan(session: SessionDep, slug: str, scan_id: uuid.UUID):
    return await scan_service.trigger_scan(session, slug, scan_id)


@router.get("/{scan_id}/jobs", response_model=list[ScanJobResponse])
async def list_scan_jobs(session: SessionDep, slug: str, scan_id: uuid.UUID):
    return await scan_service.list_scan_jobs(session, slug, scan_id)


@router.get("/{scan_id}/jobs/{job_id}", response_model=ScanJobResponse)
async def get_scan_job(
    session: SessionDep,
    slug: str,
    scan_id: uuid.UUID,
    job_id: uuid.UUID,
):
    return await scan_service.get_scan_job(session, slug, scan_id, job_id)
